src/heroesgrave/spade/core/effects/MappingEffect.java

```diff
@@ -0,0 +1,421 @@
+				g.setColor(p.b);
+				g.draw(new Ellipse2D.Float(p.a.x * getWidth() - radius, (1 - p.a.y) * getHeight() - radius, 2 * radius, 2 * radius));
+			}
+			
+			g.setStroke(tmp);
+			
+			g.setColor(Color.gray);
+			g.drawRect(0, 0, getWidth() - 1, getHeight() - 1);
+		}
+		
+		private Pair<Set<GraphEdge>, Set<Pair<Point2D.Float, Color>>> graph() {
+			Set<GraphEdge> edges = new HashSet<>();
+			Set<Pair<Point2D.Float, Color>> points = new HashSet<>();
+			boolean[] channels = new boolean[mappings.length];
+			
+			// get distinct edges and points from mappings
+			for (int i = 0; i < mappings.length; i++) {
+				if (enabled[i]) {
+					PiecewiseLinearMapping mapping = mappings[i];
+					Point2D.Float p = mapping.points.get(0);
+					points.add(new Pair<>(p, (Color) null));
+					for (int z = 1; z < mapping.points.size(); z++) {
+						Point2D.Float t = mapping.points.get(z);
+						edges.add(new GraphEdge(p, t));
+						points.add(new Pair<>(t, (Color) null));
+						p = t;
+					}
+				}
+			}
+			
+			// color points
+			for (Pair<Point2D.Float, Color> p : points) {
+				for (int x = 0; x < channels.length; x++)
+					channels[x] = enabled[x] && mappings[x].points.contains(p.a);
+				p.b = blend(channels);
+			}
+			
+			// color edges
+			for (GraphEdge e : edges) {
+				for (int i = 0; i < channels.length; i++) {
+					int aidx = mappings[i].points.indexOf(e.a);
+					int bidx = mappings[i].points.indexOf(e.b);
+					channels[i] = enabled[i] && aidx != -1 && bidx != -1 && aidx == bidx - 1;
+				}
+				
+				e.c = blend(channels);
+			}
+			
+			return new Pair<>(edges, points);
+		}
+		
+		private Color blend(boolean... channels) {
+			int v = 255;
+			for (int i = 0; i < channels.length; i++)
+				if (channels[i])
+					v -= 35;
+			
+			return new Color(channels[0] ? v : 0, channels[1] ? v : 0, channels[2] ? v : 0);
+		}
+		
+		private static class GraphEdge {
+			Point2D.Float a, b;
+			Color c;
+			
+			GraphEdge(Point2D.Float x, Point2D.Float y) {
+				a = x;
+				b = y;
+			}
+			
+			@Override
+			public boolean equals(Object e) {
+				if (e instanceof GraphEdge) {
+					GraphEdge d = (GraphEdge) e;
+					return (a.equals(d.a) && b.equals(d.b));
+				} else
+					return e.equals(this);
+			}
+		}
+	}
+	
+	// not optimized
+	private static class PiecewiseLinearMapping {
+		
+		List<Point2D.Float> points = new ArrayList<>();
+		
+		PiecewiseLinearMapping() {
+			points.add(new Point2D.Float(0, 0));
+			points.add(new Point2D.Float(1, 1));
+		}
+		
+		float get(float x) {
+			int i = 0;
+			while (i < points.size() && points.get(i).x <= x)
+				i++;
+			
+			Point2D.Float start = points.get(i - 1);
+			Point2D.Float end = points.get(Math.min(i, points.size() - 1));
+			
+			return start.y + (end.y - start.y) * ((x - start.x) / (end.x - start.x));
+		}
+		
+		static PiecewiseLinearMapping identity() {
+			return new PiecewiseLinearMapping();
+		}
+	}
+	
+	// something something tuples
+	private static class Pair<A, B> {
```

Comment:
There's already a Pair class in [`heroesgrave.utils.misc`](https://github.com/Spade-Editor/Spade/blob/master/src/heroesgrave/utils/misc/Pair.java)